### ISSUE TITLE
The Performance is completely wrong in you paper. Your level_box,py and level_feature.py are wrong 

### ISSUE THREAD

**Issue by GingerCohle** [tiankongzhang/MGAD#3]
In MGADA/FCOS/fcos_core/modeling/feature_generator/level_feature.py           Your code isn't fit your paper!!!!!!!
 bbs_iou_weights = torch.stack(bbs_iou_weights, dim=1)
            bbs_iou_weights = bbs_iou_weights - bbs_iou_weights.max(dim=1,keepdim=True)[0]/( bbs_iou_weights.max(dim=1,keepdim=True)[0] + 1e-5)
![image](https://user-images.githubusercontent.com/37873318/212835592-a088c353-7f11-4ab9-8a72-b83720b25e0e.png)

            bbs_iou_weights_list.append(F.softmax(3.8 * bbs_iou_weights, dim=1))
In MGADA/FCOS/fcos_core/data/datasets/coco.py You even change the coco code and make it multi-scale. That completely wrong. The effect you boosted came from coco multi-scale train not from your model!!!!!!!!! Without  multi-scale train,  the performance is poor OrderedDict([('bbox', OrderedDict([('AP', 0.11204973646364039), ('AP50', 0.2206766632124586), ('AP75', 0.09630345601049542), ('APs', 0.017808850902824018), ('APm', 0.11453855168008185), ('APl', 0.23813001338340156)]))]). With coco multi-scale OrderedDict([('AP', 0.20850721286259194), ('AP50', 0.36809831533667314), ('AP75', 0.20098527412543474), ('APs', 0.02997187307481019), ('APm', 0.18930275287982154), ('APl', 0.4280802012946707)]))]).
Can you update the true code!!!!!!!!!
`def __getitem__(self, idx):
        img, anno = super(COCODataset, self).__getitem__(idx)
        scale_p = npr.randint(0,10) / 10.0
        if scale_p < 0.4:
            scale_p = 1.0
        else:
            scale_p = 0.0
        scale_mark = npr.randint(0,2)*2 - 1
        scale_mark = math.pow(npr.randint(2,6), scale_mark * scale_p)
        if not self.is_sample:
            scale_p = 0.0
        #print(not self.is_sample)

        # filter crowd annotations
        # TODO might be better to add an extra field
        anno = [obj for obj in anno if obj["iscrowd"] == 0]

        boxes = [obj["bbox"] for obj in anno]

        classes = [obj["category_id"] for obj in anno]
        classes = [self.json_category_id_to_contiguous_id[c] for c in classes]
        
        ##
        im_w = img.size[0] * scale_mark
        im_h = img.size[1] * scale_mark
        im_cx = im_w / 2
        im_cy = im_h / 2

        crop_im_ex = img.size[0] / 2 - im_cx
        crop_im_ey = img.size[1] / 2 - im_cy

        rl_num_boxes = 0
        bboxes_s = []
        gt_classes_s = []
        
        if scale_p > 0:
            for idx_b in range(len(boxes)):
                bb = boxes[idx_b] * 1
                clsl = classes[idx_b]

                bb[2] = bb[0] + bb[2]
                bb[3] = bb[1] + bb[3]
                if self.domain == 0:
                    if float(bb[2]) - float(bb[0]) > 0 and float(bb[3]) - float(bb[1]) > 0:
                        bb[0] = int(bb[0] * scale_mark + crop_im_ex)
                        bb[1] = int(bb[1] * scale_mark + crop_im_ey)
                        bb[2] = int(bb[2] * scale_mark + crop_im_ex)
                        bb[3] = int(bb[3] * scale_mark + crop_im_ey)

                        if float(bb[3]) - float(bb[1]) < 16 or float(bb[2]) - float(bb[0]) < 16 or \
                            float(bb[0]) < 0 or float(bb[1]) < 0 or bb[2] > img.size[0] or bb[3] > img.size[1]:
                            continue
                        rl_num_boxes += 1
                        bboxes_s.append(bb)
                        gt_classes_s.append(clsl)
                else:
                     bb[0] = int(bb[0] * scale_mark + crop_im_ex)
                     bb[1] = int(bb[1] * scale_mark + crop_im_ey)
                     bb[2] = int(bb[2] * scale_mark + crop_im_ex)
                     bb[3] = int(bb[3] * scale_mark + crop_im_ey)

                     rl_num_boxes += 1
                     bboxes_s.append(bb)`

**Comment by GingerCohle** [comment]
The main effect comes from multi-scale training!!!!!!!!!! Please see it in 
https://github.com/chengchunhsu/EveryPixelMatters/blob/master/fcos_core/data/datasets/coco.py!!!!!!
https://github.com/CityU-AIM-Group/SIGMA/blob/main/fcos_core/data/datasets/coco.py!!!!!

**Comment by tiankongzhang** [comment]
Thank you for your question. Here are some answers to your questions.
Q1. Inconsistency of formula？
For modern version of code,  method of weight generation is consistent with Eq.(3) except for an extra operation to divide by the maximum weight. You can also remove it.
Q2. In our paper, we illustrate the use of data enhancement(scale change), i.e, Implementation Details
Q3. all of data enhancement, our Gate and multi-Granularity Alignment method can enhance the performance of detector. In paper, we mainly show the effectiveness of our model and method, i.e, Table. 4
Q4. There are some different between the current version and the final version of our paper. Due to graduation during this time, I failed to update in time. I will update the full version and extended version as soon as possible.

**Comment by GingerCohle** [comment]
Thanks for yr reply! The following formula is equal to bbs_iou_weights-1
bbs_iou_weights = bbs_iou_weights - bbs_iou_weights.max(dim=1,keepdim=True)[0]/( bbs_iou_weights.max(dim=1,keepdim=True)[0] + 1e-5). 
The multi-scale strategy contributes too much to the model performance which we usually illustrate in the experiment table. When I trained it with tesla v100, the ms scale training boosted from 0.2206766632124586 to 0.36809831533667314 with respect to AP50. The papers you compared are based on single-scale training, which is unfair. In the field of OD, multi-scale training is a trick and we always illustrate it in the experiment table. If someone, like me, wants to follow your work, the result is frustrating.
Waiting for you to update the latest code! Ty!
